Assert that kernels are called with the right signature

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39611 Unify TensorOptions signatures
* #39492 Move TensorOptions ops to c10
* **#38361 Assert that kernels are called with the right signature**

Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

**Update:** It's not in the call path anymore but it's now in the path when you query an OperatorHandle.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

smessmer committed:
Assert that kernels are called with the right signature

Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/boxing/KernelFunction_impl.h:
```diff
@@ -50,6 +62,13 @@ inline Return KernelFunction::call(const OperatorHandle& opHandle, Args... args)
     // forwarding, which would require Args to be deduced, but instead we
     // want callers to explicitly specify the Args.
 
+    TORCH_INTERNAL_ASSERT(!signature_hash_.has_value() || (detail::hashFunctionSignature<Return (Args...)>() == *signature_hash_),
```
I've said this before and I'll say it again: I don't want an extra conditional test in the hotpath for all of our unboxed calls. If you bake the unboxed calling type into UnboxedHandle you can do the check once when we do the lookup (which is slow path anyway)

smessmer replied:
fixed

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -101,6 +102,17 @@ class CAFFE2_API OperatorEntry final {
+        "Tried to access operator ", name_, " with a wrong signature. Accessed with ",
+        CppSignature::make<FuncType>().name(),
+        " but the operator was registered with ",
+        cpp_signature_->name()
```
I'd suggest adding some suggestions about how to fix the problem here

ezyang commented on aten/src/ATen/core/dispatch/CppSignature.h:
```diff
@@ -0,0 +1,46 @@
+        // Normalize functors, lambdas, function pointers, etc. into the plain function type
+        using decayed_function_type = typename guts::infer_function_traits_t<std::decay_t<FuncType>>::func_type;
+
+        return CppSignature(std::type_index(typeid(decayed_function_type)));
```
Use of typeid here means that you must compile with RTTI for this to work. We have some settings where we don't compile with RTTI. What happens in those cases?

smessmer replied:
we don't have non-RTTI builds atm

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -279,6 +278,66 @@ class CAFFE2_API OperatorHandle final {
+ * unboxed way.
+ */
+template<class FuncType>
+class CAFFE2_API TypedOperatorHandle final {
```
Suggestion: have this inherit publicly from OperatorHandle, and then just add the typed caller as a new method. This would save you a bunch of boilerplate. I know the modern OO mantra is "prefer composition over inheritance", but TypedOperatorHandle truly fulfills the "is-a" relationship here!

ezyang commented:
The test failures might be because in some situations, we have multiple binary compatible versions of the library which have different RTTI. I've noticed this previously in ASAN. Here is a relevant note from the CI:

```
      # NB: We load libtorch.so with RTLD_GLOBAL for UBSAN, unlike our
      # default behavior.
      #
      # The reason for this is that without RTLD_GLOBAL, if we load multiple
      # libraries that depend on libtorch (as is the case with C++ extensions), we
      # will get multiple copies of libtorch in our address space.  When UBSAN is
      # turned on, it will do a bunch of virtual pointer consistency checks which
      # won't work correctly.  When this happens, you get a violation like:
      #
      #    member call on address XXXXXX which does not point to an object of
      #    type 'std::_Sp_counted_base<__gnu_cxx::_Lock_policy::_S_atomic>'
      #    XXXXXX note: object is of type
      #    'std::_Sp_counted_ptr<torch::nn::LinearImpl*, (__gnu_cxx::_Lock_policy)2>'
      #
      # (NB: the textual types of the objects here are misleading, because
      # they actually line up; it just so happens that there's two copies
      # of the type info floating around in the address space, so they
      # don't pointer compare equal.  See also
      #   https://github.com/google/sanitizers/issues/1175
      #
      # UBSAN is kind of right here: if we relied on RTTI across C++ extension
      # modules they would indeed do the wrong thing;  but in our codebase, we
      # don't use RTTI (because it doesn't work in mobile).  To appease
      # UBSAN, however, it's better if we ensure all the copies agree!
      #
      # By the way, an earlier version of this code attempted to load
      # libtorch_python.so with LD_PRELOAD, which has a similar effect of causing
      # it to be loaded globally.  This isn't really a good idea though, because
      # it depends on a ton of dynamic libraries that most programs aren't gonna
      # have, and it applies to child processes.
```

One way I know how to fix this problem is if we roll our own RTTI mechanism, manually assembling a type index. But that's a bunch of work. I'm not sure if there's any better option. Resolving the RTLD_GLOBAL problem will be quite challenging in general.

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -232,7 +235,7 @@ class CAFFE2_API Dispatcher final {
  * This handle can be used to register kernels with the dispatcher or
  * to lookup a kernel for a certain set of arguments.
  */
-class CAFFE2_API OperatorHandle final {
```
RIP final

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -101,6 +102,18 @@ class CAFFE2_API OperatorEntry final {
+        CppSignature::make<FuncType>().name(),
+        " but the operator was registered with ",
+        cpp_signature_->name(),
+        " This likely happened in a call to OperatorHandle::typed<Return (Args...)>(). Please make sure that the function signature matches the signature in the operator registration call."
```
Can you print out the `debug_` info here? That will let people know where the operator was registered

ezyang reviewed (approved):
yay, much needed!

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]

smessmer commented:
> The test failures might be because in some situations, we have multiple binary compatible versions of the library which have different RTTI. I've noticed this previously in ASAN. Here is a relevant note from the CI:
> 
> ```
>       # NB: We load libtorch.so with RTLD_GLOBAL for UBSAN, unlike our
>       # default behavior.
>       #
>       # The reason for this is that without RTLD_GLOBAL, if we load multiple
>       # libraries that depend on libtorch (as is the case with C++ extensions), we
>       # will get multiple copies of libtorch in our address space.  When UBSAN is
>       # turned on, it will do a bunch of virtual pointer consistency checks which
>       # won't work correctly.  When this happens, you get a violation like:
>       #
>       #    member call on address XXXXXX which does not point to an object of
>       #    type 'std::_Sp_counted_base<__gnu_cxx::_Lock_policy::_S_atomic>'
>       #    XXXXXX note: object is of type
>       #    'std::_Sp_counted_ptr<torch::nn::LinearImpl*, (__gnu_cxx::_Lock_policy)2>'
>       #
>       # (NB: the textual types of the objects here are misleading, because
>       # they actually line up; it just so happens that there's two copies
>       # of the type info floating around in the address space, so they
>       # don't pointer compare equal.  See also
>       #   https://github.com/google/sanitizers/issues/1175
>       #
>       # UBSAN is kind of right here: if we relied on RTTI across C++ extension
>       # modules they would indeed do the wrong thing;  but in our codebase, we
>       # don't use RTTI (because it doesn't work in mobile).  To appease
>       # UBSAN, however, it's better if we ensure all the copies agree!
>       #
>       # By the way, an earlier version of this code attempted to load
>       # libtorch_python.so with LD_PRELOAD, which has a similar effect of causing
>       # it to be loaded globally.  This isn't really a good idea though, because
>       # it depends on a ton of dynamic libraries that most programs aren't gonna
>       # have, and it applies to child processes.
> ```
> 
> One way I know how to fix this problem is if we roll our own RTTI mechanism, manually assembling a type index. But that's a bunch of work. I'm not sure if there's any better option. Resolving the RTLD_GLOBAL problem will be quite challenging in general.

CI on the PR is passing now. I'll try landing this and see what happens. If something breaks in the master-only test cases, we might have to revert this.

smessmer committed:
Update on "Assert that kernels are called with the right signature"


Rather than segfaulting, we should show a good error message when in op.call<Return, Args...>(...) the Return type or Args types mismatch the kernel.

This adds an assertion comparing two std::type_index to the call path, but that should be fast. Hashing the function signature is also in the call path and not strictly constexpr, but I checked on godbolt that GCC >=5 and Clang >=3.8 optimize it away and make it constexpr, i.e. it's not part of the assembly.

supersedes https://github.com/pytorch/pytorch/pull/26482

Differential Revision: [D21534052](https://our.internmc.facebook.com/intern/diff/D21534052/)

[ghstack-poisoned]
